Documentation for neural nets in MindMeld (#393)
cisco/mindmeld: source/userguide/deep_neural_nets.rst

Deep Neural Networks in MindMeld
================================

The advent of `Deep learning <https://en.wikipedia.org/wiki/Deep_learning>`_ has led to substantial developments in the area of Conversational AI as well as Natural Langauge Processing (NLP) in general.
Specifically, the proliferation of deep neural models based on :wiki_api:`Convolutional neural networks (CNN) <Convolutional_neural_network#Natural_language_processing>`, :wiki_api:`Long short-term memory networks (LSTM) <Long_short-term_memory>` and :wiki_api:`Transformers <Transformer_(machine_learning_model)>`, and their rapid adaptation to wide range of NLP tasks has been a great success.

> Thus extending the suite of traditional machine learning models (e.g. :sk_guide:`Logistic regression <linear_model.html#logistic-regression>`, :sk_guide:`Decision tree <tree.html#tree>`, etc.), MindMeld provides the option for users to train and use a variety of deep neural models with an array of configurable parameters.

Reviewer: ```suggestion
MindMeld extends its suite of traditional machine learning models (e.g. :sk_guide:`Logistic regression <linear_model.html#logistic-regression>`, :sk_guide:`Decision tree <tree.html#tree>`, etc.) with a variety of deep neural models with an array of configurable parameters.
```

Author: Related changes made in [Kelsey's PR](https://github.com/cisco/mindmeld/blob/feature/doc-review-nn_support_phase2_mm_docs/source/userguide/deep_neural_nets.rst)

> The deep neural models are available both for text classification aka. sequence classification (:ref:`domain classification <domain_classification>` and :ref:`intent classification <intent_classification>`) as well as for token classification (:ref:`entity recognition <entity_recognition>`) tasks.

Author: Related changes made in [Kelsey's PR](https://github.com/cisco/mindmeld/blob/feature/doc-review-nn_support_phase2_mm_docs/source/userguide/deep_neural_nets.rst)

> The models are implemented using `Pytorch <https://pytorch.org/>`_ framework and thus need extra installations before starting to use them in your chatbot application.

Reviewer: Think you can move this to the note?

Author: Related changes made in [Kelsey's PR](https://github.com/cisco/mindmeld/blob/feature/doc-review-nn_support_phase2_mm_docs/source/userguide/deep_neural_nets.rst)

.. note::

   Please make sure to install the Pytorch requirement by running in the shell: :code:`pip install mindmeld[torch]`.

MindMeld supports the use of pretrained transformer models such as :wiki_api:`BERT <BERT_(language_model)>` through the popular `Huggingface Transformers <https://huggingface.co/docs/transformers/index>`_ library.

> Several pretrained models from their `Models Hub <https://huggingface.co/models>`_ that can be used for sequence classification or token classification can be employed in your chatbot application.

Reviewer: ```suggestion
Several pretrained models from their `Models Hub <https://huggingface.co/models>`_ can be used for sequence or token classification to be employed in your chatbot application.
```

.. note::

   Install the extra *transformers* requirement by running in the shell: :code:`pip install mindmeld[transformers]`.

Before proceeding to use the deep neural models, consider the following possible advantages and disadvantages of using them in place of traditional machine learning models.

Some advantages are:

> - The deep models generally outperform traditional machine learning models on training sets with at least several hundreds or few thousands of queries when training from scratch, and with at least few hundreds if fine-tuning from a pretrained checkpoint.

Reviewer: ```suggestion
- The deep models generally outperform traditional machine learning models on training sets with several hundreds or few thousands of queries when training from scratch, and with at least few hundreds if fine-tuning from a pretrained checkpoint.
```

Reviewer: Actually, not sure this is an advantage or a disadvantage. Depends on the amount of data a user has.

- The deep models require little or no feature engineering work than traditional machine learning models, because they learn the input features (such as word embeddings) compared to several hundred engineered features (n-grams, system entities, and so on) of the latter.
- On GPU-enabled devices, the deep networks can achieve training time comparable to some of the traditional models in MindMeld.

Some disadvantages are:

- Training time on CPU-only machines is a lot slower than for traditional machine learning models.
- No automated hyperparameter tuning methods like :sk_api:`sklearn.model_selection.GridSearchCV <sklearn.model_selection.GridSearchCV.html>` are available for deep neural models.
- The deep neural models generally occupy similar or more disk storage space compared to their traditional counterparts.

Parameter tuning for deep neural models is more involved than for traditional machine learning models.

> A good starting point for understanding this subject is Andrej Karpathy's `course notes <https://cs231n.github.io/neural-networks-3/#baby>`_ from the Convolutional Neural Networks for Visual Recognition course at Stanford University.

Reviewer: Say something like - While we have good defaults for all params that works well across datasets, you can understand this better using this course.

.. note::

>    To use deep neural networks instead of traditional machine learning models, simply make few modifications to the classifier configuration dictionaries for all or selected classifiers in your app's ``config.py``.

Reviewer: If we're not mentioning what these changes are or linking to a page that mentions them, this note seems unnecessary?

In the following sections, different model architectures and their configurable parameters are outlined.

Domain and Intent classification
--------------------------------

.. _dnns_sequence_classification:

Recall from :ref:`Working with the Domain Classifier <domain_classifier_configuration>` and :ref:`Working with the Intent Classifier <intent_classifier_configuration>` sections that a text classifier configuration consists of the keys

- ``'features'``,
- ``'param_selection'``,
- ``'model_settings'``, and
- ``'params'``

amongst other keys that do not have distinction between traditional models or deep neural models.

> When working with deep neural models, the ``'features'`` and ``'param_selection'`` keys in the classifier configuration are redundant as we neither have to handcraft any feature sets for modeling nor there is an automated hyperparameter tuning.

Reviewer: ```suggestion
When working with deep neural models, the only relevant keys to be configured are ``'model_settings'`` and ``'params'``. We do not need the``'features'`` and ``'param_selection'`` keys in the classifier configuration we don't use handcrafted feature sets and no automated hyperparameter tuning.
```

In its typical architecture for entity recognition, the first layer embeds the sequence of textual tokens obtained from input text into low-dimensional vectors using an embedding lookup table.
The capacity of an LSTM in maintaining the temporal information is generally dependant on its *hidden* dimension.
Further, several LSTM layers can be stacked one-after-another and each layer can process the text from just beginning-to-end or both ways.
The first layer's output embedding sequence is passed through the stacked LSTMs, which finally produces one vector per token of the input text.

To obtain a single vector per word per input text, the vectors of all tokens corresponding to each word (for which an entity tag is to be ascertained) are pooled.
This is unlike sequence classification models where all tokens of all words are pooled together.
The pooled outputs are then passed through a classification layer.
Dropout layers are used as regularizers to avoid over-fitting, which is a more common phenomenon when working with small sized datasets.

> Following are the different optional params that are configurable with the ``'lstm'`` classifier type.

Reviewer 2: `'lstm-pytorch'` classifier type?

concatenated with outputs of convolutions at character-level (for each word individually) using kernels of different lengths to capture different patterns.
The capacity of an LSTM in maintaining the temporal information is generally dependant on its *hidden* dimension.
Further, several LSTM layers can be stacked one-after-another and each layer can process the text from just beginning-to-end or both ways.
The first layer's output embedding sequence is passed through the stacked LSTMs, which finally produces one vector per token of the input text.

To obtain a single vector per word per input text, the vectors of all tokens corresponding to each word (for which an entity tag is to be ascertained) are pooled.
This is unlike sequence classification models where all tokens of all words are pooled together.
The pooled outputs are then passed through a classification layer.
Dropout layers are used as regularizers to avoid over-fitting, which is a more common phenomenon when working with small sized datasets.

> Following are the different optional params that are configurable with the ``'lstm'`` classifier type.

Reviewer 2: `'cnn-lstm'` classifier type?

Below is a minimalistic example of a token classifier configuration for classifier based on CNN-LSTM:

.. code-block:: python

   {
    'model_type': 'tagger',
    'train_label_set': 'train.*\.txt',
    'test_label_set': 'test.*\.txt',
    'model_settings': {'classifier_type': 'cnn-lstm'},
    'params': {
        'embedder_type': 'glove',
        'lstm_hidden_dim': 128,
        'lstm_bidirectional': True,
        'char_emb_dim': 32
    },
   }

4. ``'lstm-lstm'`` classifier type
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. _dnns_token_classification_models_lstm_lstm:

:wiki_api:`Long short-term memory networks (LSTM) <Long_short-term_memory>` based text classifiers are classifiers that utilize recurrent feedback connections to be able to learn temporal dependencies in sequential data.
When coupled with :wiki_api:`Long short-term memory networks (LSTM) <Long_short-term_memory>` for extracting character-level features from input text, the overall architecture can better model the textual data as well as it is more robust to variations in the spellings.

> In its typical architecture for entity recognition, the first layer embeds the sequence of words obtained f

Reviewer 2: `'lstm-lstm'` classifier type?

> In its typical architecture for entity recognition, the first layer embeds the sequence of words obtained f

Reviewer 2: This section is structured differently than the other 4 sections of this type, in that it has a straightforward bulleted list of the key points for this section. The other sections do a good job of laying out the key points, too, but they do so in paragraph form. 

I'm kind of partial to this display of information, though it is a bit of a departure from what appears to be the norm for MindMeld documentation. On the one hand, it might be a lot of work to rewrite the other sections in this 'Entity recognition'  section as opposed to making this conform to the style imposed in those other sections. But on the other hand, this is a really nice point-for-point outline of what is happening with this classifier type, and it's presented in a very user-friendly visual and organizational style.
